### PR TITLE
Add reverseproxy auth for API back with default disabled

### PR DESCRIPTION
**Pull request by lunny**
This feature was removed by #22219 to avoid possible CSRF attack.

This PR takes reverseproxy auth for API back but with default disabled.

To prevent possbile CSRF attack, the responsibility will be the reverseproxy but not Gitea itself.

For those want to enable this `ENABLE_REVERSE_PROXY_AUTHENTICATION_API`, they should know what they are doing.

### CODE REVIEW

**Comment by techknowlogick** [comment]
cc: @pboguslawski as this is likely relevant to your interests

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -621,7 +621,8 @@ And the following unique queues:
    BASIC and the user's password. Please note if you disable this you will not be able to access the
    tokens API endpoints using a password. Further, this only disables BASIC authentication using the
    password - not tokens or OAuth Basic.
-- `ENABLE_REVERSE_PROXY_AUTHENTICATION`: **false**: Enable this to allow reverse proxy authentication.
+- `ENABLE_REVERSE_PROXY_AUTHENTICATION`: **false**: Enable this to allow reverse proxy authentication for web requests
+- `ENABLE_REVERSE_PROXY_AUTHENTICATION_API`: **false**: Enable this to allow reverse proxy authentication for API requests
```
Comment (silverwind) [inline docs/content/administration/config-cheat-sheet.en-us.md]: Don't forget to update `custom/conf/app.example.ini`

**Review by silverwind — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]
LGTM. But I would like the author and reviewers of #22219 (@zeripath @jolheiser @delvh) to take a look.

File: docs/content/administration/config-cheat-sheet.en-us.md (hunk 1 above)
Comment (delvh) [inline docs/content/administration/config-cheat-sheet.en-us.md]: Perhaps we should explicitly mention that the reverse proxy is responsible for ensuring that no CSRF is possible.

Comment (lunny) [reply docs/content/administration/config-cheat-sheet.en-us.md]: resolved.

**Review by jolheiser — APPROVED** [review]
I don't love it, but I'll concede that some users relied on it and there may not be an ideal alternative yet.

**Review by delvh — APPROVED** [review]

**Comment by lunny** [comment]
I think we can continue since @zeripath has no response for 3 days.

**Pull request merged**